core/services/flux_monitor.go

```diff
@@ -0,0 +1,346 @@
+}
+
+type addEntry struct {
+	job   *models.JobSpec
+	rchan chan error
+}
+
+// NewFluxMonitor creates a service that manages a collection of DeviationCheckers,
+// one per initiator of type InitiatorFluxMonitor for added jobs.
+func NewFluxMonitor(store *store.Store, runManager RunManager) FluxMonitor {
+	return &concreteFluxMonitor{
+		store:          store,
+		runManager:     runManager,
+		checkerFactory: pollingDeviationCheckerFactory{},
+	}
+}
+
+func (fm *concreteFluxMonitor) Start() error {
+	fm.ctx, fm.cancel = context.WithCancel(context.Background())
+	fm.adds = make(chan addEntry)
+	fm.removes = make(chan *models.ID)
+	fm.connect = make(chan struct{})
+	fm.disconnect = make(chan struct{})
+
+	go fm.actionConsumer(fm.ctx) // start single goroutine consumer
+
+	rchan := make(chan error, 1)
+	count := 0
+	err := fm.store.Jobs(func(j *models.JobSpec) bool { // add persisted jobs
+		fm.adds <- addEntry{j, rchan}
+		count++
+		return true
+	}, models.InitiatorFluxMonitor)
+
+	// Block until jobs have been added, returning errors if any.
+	var merr error
+	for i := 0; i < count; i++ {
+		merr = multierr.Combine(merr, <-rchan)
+	}
+	return multierr.Append(err, merr)
+}
+
+// Connect initializes all DeviationCheckers and starts their listening.
+func (fm *concreteFluxMonitor) Connect(*models.Head) error {
+	fm.connect <- struct{}{}
+	return nil
+}
+
+// actionConsumer is the CSP consumer. It's run on a single goroutine to
+// coordinate the collection of DeviationCheckers in a thread-safe fashion.
+func (fm *concreteFluxMonitor) actionConsumer(ctx context.Context) {
+	jobMap := map[string][]DeviationChecker{}
+
+	// init w a noop cancel, so we never have to deal with nils
+	connectionCtx, cancelConnection := context.WithCancel(ctx)
+	var connected bool
+
+	for {
+		select {
+		case <-ctx.Done():
+			cancelConnection()
+			return
+		case <-fm.connect:
+			// every connection, create a new ctx for canceling on disconnect.
+			connectionCtx, cancelConnection = context.WithCancel(ctx)
+			connectCheckers(connectionCtx, jobMap, fm.store.TxManager)
+			connected = true
+		case <-fm.disconnect:
+			cancelConnection()
+			connected = false
+		case entry := <-fm.adds:
+			entry.rchan <- fm.addAction(connectionCtx, connected, entry.job, jobMap)
+		case jobID := <-fm.removes:
+			for _, checker := range jobMap[jobID.String()] {
+				checker.Stop()
+			}
+			delete(jobMap, jobID.String())
+		}
+	}
+}
+
+// Disconnect cleans up running deviation checkers.
+func (fm *concreteFluxMonitor) Disconnect() {
+	fm.disconnect <- struct{}{}
+}
+
+// Disconnect cleans up running deviation checkers.
+func (fm *concreteFluxMonitor) Stop() {
+	fm.cancel()
+}
+
+// OnNewHead is a noop.
+func (fm *concreteFluxMonitor) OnNewHead(*models.Head) {}
+
+// AddJob created a DeviationChecker for any job initiators of type
+// InitiatorFluxMonitor.
+func (fm *concreteFluxMonitor) AddJob(job models.JobSpec) error {
+	if !job.IsFluxMonitorInitiated() {
+		return nil
```

Comment:
Shouldn't this be an error? Should this be validated earlier?

Reply:
We have a precedence where it's a noop if you try to add a job that doesn't have a matching initiator:

```
func (app *ChainlinkApplication) AddJob(job models.JobSpec) error {
	err := app.Store.CreateJob(&job)
	if err != nil {
		return err
	}

	app.Scheduler.AddJob(job)

	// XXX: Add mechanism to asynchronously communicate when a job spec has
	// an ethereum interaction error.
	// https://www.pivotaltracker.com/story/show/170349568
	logger.ErrorIf(app.FluxMonitor.AddJob(job))
	logger.ErrorIf(app.JobSubscriber.AddJob(job, nil))
	return nil
}
```

I'm fine with it as is, having the check for matching initiators in the initiator. It has some nuance to it, like jobSubscriber looks for models.InitiatorEthLog, models.InitiatorRunLog, models.InitiatorServiceAgreementExecutionLog

Reply:
When would it ever happen? I guess I worry that if it did happen, it's masking some heinous shit.

Reply:
When would what happen exactly? Want to talk specifics in slack?